### ISSUE TITLE
Workflows with action policies can deadlock

### ISSUE THREAD

**Issue by bri365** [StackStorm/st2#2814]
Given the attached pack (useful for meditation), the system deadlocks when the action _dispatcher_pool (50 entries) gets filled with workflows, and all workflow actions are subsequently held in the _work_buffer queue indefinitely.

One possible solution would be to limit the number of workflows in the _dispatcher_pool to one less than half the total entries, in order to leave room for action execution.

[deadlock.tar.gz](https://github.com/StackStorm/st2/files/361933/deadlock.tar.gz)

# Deadlock pack

This pack contains actions for testing policy deadlock and other behaviors

## Actions

* ``delay2`` - a two second delay
* ``delay20`` - a twenty second delay
* ``delay2test`` - a workflow with delay2 followed by delay20

## Policies

* ``delay20`` - limit concurrency to 1
* ``delay2test`` - limit concurrency to 10

## Testing

Launch multiple instances of the workflow:

* ``for i in `seq 1 100`; do st2 action execute deadlock.delay2test; done``

Without the ``delay2test`` policy, the actions deadlock

**Comment by bri365** [comment]
Another, potentially better, solution would be to have separate queues for workflows and actions, so actions can be protected from a queue full of workflows.  Even better, three queues - workflows, workflow actions, and regular actions; now workflow actions are not subjected to the average queue depth of the entire action list.

**Comment by Kami** [comment]
Thanks for including he pack which helps to reproduce the issue - this should help us a lot tracking down the root cause.

I will have a look when I get a chance.

I do think that having multiple internal queues / pools sounds like a good idea, but I need to dig in and think about it some more :)

**Comment by Kami** [comment]
I personally couldn't reproduce the deadlock issue with the provided code (even with a smaller pool with a size of 2 so we hit the issue earlier).

I used a single action runner and didn't register `delay2test` policy (with multiple action runners it should be even less of an issue because we have a pool per action runner so `50 * number of action runners` should be a maximum concurrent throughput).

I do see the general problem though - the problem is that dispatcher will be blocked when all the threads are occupied. This won't cause a deadlock (at least I couldn't reproduce it), but it will cause a temporary blockage / delay until the pool frees up. In theory things work as intended / designed - we limit the pool size because we want to prevent resource starvation.

I do agree that this is not ideal in all scenarios though. For one we could at least allow user to configure the maximum dispatcher pool size. As mentioned above, we do need some kind of upper limit to prevent the resource starvation though - depending on the server specs this could higher than a default value of 50.

So for one I propose a config change which allows operator to specify a maximum dispatcher thread pool size for each action runner. I still think 50 is a reasonable default for now, but the actual maximum value really depends on each StackStorm installation and workload (number of concurrent actions with delay, type of actions - long running CPU intensive action vs long running on CPU intensive actions, etc.)

**Comment by Kami** [comment]
In any case, I'm also looking into having multiple Dispatchers per different message type (as suggested above).

**Comment by Kami** [comment]
@lakshmi-kannan Please check the discussion above and my comment below and let me know what you think.

I propose the following change:

1. Introduce a new QueueConsumer class for action runner.
2. This QueueConsumer uses multiple `BufferedDispatcher` green thread pools - one for regular actions and one for workflows.
3. Default to 50, but allow operator to configure / tweak maximum pool size for both of those pools.

I think this should work reasonably well for the problem described above and it's also similar to what @bri365 has suggested.

If we go with this approach there is one problem with the implementation though - right now the handler and consumer don't know if a particular message represents a regular action or workflow action execution. If we want to determine that, we need to retrieve `ActionDB` object for the `LiveActionDB` message which is currently being processed.

This means 1 GET by primary key query for each message before we dispatch it to the dispatcher pool. This might not be a big deal since it's "get by PK", but I would still prefer to avoid this query, if possible (for one, we could amend `LiveActionDB` object to carry this information with it).

What do others think?

**Comment by bri365** [comment]
@Kami - sorry, you have to delete the policy delay2test.yaml before registering the pack; I used it to work around the issue.  it regulates the workflows and doesn't let them fill the work buffer

**Comment by bri365** [comment]
I like the three proposals above; they should do the trick.  
+1 for adding a workflow flag to LiveActionDB objects.

**Comment by Kami** [comment]
@bri365 Thanks for the feedback.

I will also wait on @lakshmi-kannan and others to chime in and if they agree with my proposed approach, I will go ahead and implement it :)

**Comment by lakshmi-kannan** [comment]
So far we've tried hard to make sure we don't differentiate actions from workflows and we've been successful. I still don't see a clear need for separating actions from workflows in terms of execution. I am fine with a configurable parameter for the dispatcher pool and this is something we've discussed in the past but I definitely don't see a need for two different queues. I'll hold off on implementing it. If you used mistral as the workflow engine, the eventlet basically shoots off a HTTP message to mistral API and then is pretty much done. The only case where a workflow can compete with actions is when there is an action chain and a runner picked up an action chain to execute. Even in this case, even with one runner, since we use eventlets, I think we'll be fine. So in either case, we don't need a separate queue for workflows. @Kami Can we do a hangout so we are on same page first?

**Comment by Kami** [comment]
@lakshmi-kannan Probably best way for you to understand it is to check the code and play with the example @bri365 provided.

**Comment by Kami** [comment]
Also, I would be OK to begin with by just making the pool size operator configurable.

Having said that, we already differentiate actions and workflows internally and this change would be internal as well (it wouldn't be user exposed, it's an implementation detail aka performance optimization and usually when you do optimizations like that you can't keep them 100% "generic" and need some additional details) so I personally don't see implementing two dispatcher pools / queues as an issue.

**Comment by lakshmi-kannan** [comment]
To clarify, I think I see the problem - the dispatcher queue size can be outrun by the number of workflows and actions. This means you are essentially hitting the scaling limit. The two queue solution kami is proposing only partly alleviates the problem but you can construct a case there where you can hit capacity given enough number of workflows and actions. So I don't necessarily think having two queues is better than one. It seems like you'll need monitoring to check the rabbit queue size and capacity planning up front. @Kami and I seem to disagree on the value of two queues. So either I need more convincing or other engineers have to break our tie :).

**Comment by Kami** [comment]
Yes, as discussed offline, this won't fully resolve the issue (nothing will, we need hard limits anyway), but it will increase the safety margin and will make hitting this issue less likely.

It's also worth noting that the actual setting is very workload / use-case and hardware specific. In some cases, a pool size of 50 might be too much already and in other scenarios it might be way too low. In addition to that we of course also need to have some hard limit (which is configuration and workload specific) to avoid resource exhaustion and blocking because of it.

Having said that, I do think having two pools has benefits (and is better than having one). It means workflows won't block workflow actions which will increase the safety margins further and make it even less likely that this issue will be hit (right now you can have a scenario where workloads itself block workflow actions aka deadlock, with two queues you can still hit it but much less likely since workflow actions will still be executed even if the workflow queue is full).

So in short, I think we should have two pools with some reasonable default size (50 for each or similar) and allow user to tune it (aka as proposed and described above). User will of course need to have monitoring in place (nothing new or unexpected) to determine what are the optimal sizes for their use case and hardware (optimal = where resources are utilized well, but not overloaded and where queue sizes and large enough to prevent exhaustion / deadlock from happening before hardware resource limits are hit).

And again, TL:DR the main problem is that right now deadlock can be hit before server / hardware resources are exhausted and that's obviously not ideal.

**Comment by bri365** [comment]
Reconfirmed issue with action chains on a two core 8GB system (defaults to two actionrunners):
root@stackstorm:~# history
    1  apt-get update
    2  apt-get upgrade
    3  curl -sSL https://stackstorm.com/packages/install.sh | bash -s -- --user=st2admin --password=password
    4  cd /opt/stackstorm/packs/
    5  tar -xvzf /home/brian/deadlock4.tar.gz
    6  st2 run packs.setup_virtualenv packs=deadlock
    7  st2 run packs.load register=all

root@stackstorm:/opt/stackstorm/packs/deadlock# cat chain-deadlock.sh
for i in `seq 1 200`; do st2 action execute deadlock.delay2test; done
(twice as many as total _work_buffer slots)

root@stackstorm:/opt/stackstorm/packs/deadlock# ./chain-deadlock.sh
To get the results, execute:
 st2 execution get 5790290c02ebd504749035c6
To get the results, execute:
 st2 execution get 5790290c02ebd504749035c9
...
To get the results, execute:
 st2 execution get 5790295b02ebd5047490381b

root@stackstorm:/opt/stackstorm/packs/deadlock# mongo
> use st2
switched to db st2
> db.action_execution_d_b.count({status:"delayed"})
88
> db.action_execution_d_b.count({status:"scheduled"})
109
> db.action_execution_d_b.count({status:"running"})
100
> exit
bye
root@stackstorm:/opt/stackstorm/packs/deadlock# date
Wed Jul 20 21:50:00 EDT 2016
root@stackstorm:/opt/stackstorm/packs/deadlock# mongo
MongoDB shell version: 2.4.9
connecting to: test
> use st2
switched to db st2
> db.action_execution_d_b.count({status:"delayed"})
88
> db.action_execution_d_b.count({status:"scheduled"})
109
> db.action_execution_d_b.count({status:"running"})
100
> exit
bye
root@stackstorm:/opt/stackstorm/packs/deadlock# date
Wed Jul 20 22:02:07 EDT 2016
root@stackstorm:/opt/stackstorm/packs/deadlock# mongo
MongoDB shell version: 2.4.9
connecting to: test
> use st2
switched to db st2
> db.action_execution_d_b.count({status:"delayed"})
88
> db.action_execution_d_b.count({status:"scheduled"})
109
> db.action_execution_d_b.count({status:"running"})
100
> exit
bye
root@stackstorm:/opt/stackstorm/packs/deadlock# date
Wed Jul 20 22:15:37 EDT 2016
root@stackstorm:/opt/stackstorm/packs/deadlock# mongo
MongoDB shell version: 2.4.9
connecting to: test
> use st2
switched to db st2
> db.action_execution_d_b.count({status:"delayed"})
88
> db.action_execution_d_b.count({status:"scheduled"})
109
> db.action_execution_d_b.count({status:"running"})
100
> exit
bye

[deadlock4.tar.gz](https://github.com/StackStorm/st2/files/375058/deadlock4.tar.gz)

**Comment by Kami** [comment]
@bri365 Thanks for the report.

I'm working on the change mentioned above - should be ready for review by end of this week :)

**Comment by bri365** [comment]
@Kami Thanks!  Mistral workflows, by the way, seem to be unaffected - I queued up 400 on the same system and they are dutifully chunking through, one per action runner, 20 seconds at a time.

**Comment by Kami** [comment]
@bri365 You are welcome.

And yeah, Mistral runner is mostly just "fire and forget" (when scheduling / running an execution, it just sends an HTTP request to the Mistral API).

**Comment by Kami** [comment]
Here we go - #2823

**Comment by Kami** [comment]
@bri365 Alright, #2823 has been merged into master. Thanks again for reporting this issue and including code to reproduce it.

If you feel a bit adventurous, you can try v1.6dev packages once they are built (IIRC, there are no backward incompatible changes in v1.6 so it should be fine).

**Comment by kestrel5** [comment]
@kami will do; thanks!

**Comment by bri365** [comment]
@Kami works like a charm - thanks!  I do see some unexpected order of operations, though:
(virtualenv)brian@ubuntu1404:~/st2$ st2 execution list -n 0 --status succeeded
+----------------------------+------------------------+--------------+--------------------------+-------------------------------+-------------------------------+
| id                         | action.ref             | context.user | status                   | start_timestamp               | end_timestamp                 |
+----------------------------+------------------------+--------------+--------------------------+-------------------------------+-------------------------------+
|   5797c81c0e77764649b3882f | packs.setup_virtualenv | stanley      | succeeded (5s elapsed)   | Tue, 26 Jul 2016 20:29:16 UTC | Tue, 26 Jul 2016 20:29:21 UTC |
| + 5797c82d0e77764649b38832 | deadlock.delay2test    | stanley      | succeeded (26s elapsed)  | Tue, 26 Jul 2016 20:29:33 UTC | Tue, 26 Jul 2016 20:29:59 UTC |
| + 5797c8520e77764649b38835 | deadlock.delay2test    | stanley      | succeeded (36s elapsed)  | Tue, 26 Jul 2016 20:30:10 UTC | Tue, 26 Jul 2016 20:30:46 UTC |
| + 5797c8530e77764649b38838 | deadlock.delay2test    | stanley      | succeeded (96s elapsed)  | Tue, 26 Jul 2016 20:30:11 UTC | Tue, 26 Jul 2016 20:31:47 UTC |
| + 5797c8530e77764649b3883b | deadlock.delay2test    | stanley      | succeeded (117s elapsed) | Tue, 26 Jul 2016 20:30:11 UTC | Tue, 26 Jul 2016 20:32:08 UTC |
| + 5797c8540e77764649b3883e | deadlock.delay2test    | stanley      | succeeded (137s elapsed) | Tue, 26 Jul 2016 20:30:12 UTC | Tue, 26 Jul 2016 20:32:29 UTC |
| + 5797c8540e77764649b38841 | deadlock.delay2test    | stanley      | succeeded (159s elapsed) | Tue, 26 Jul 2016 20:30:12 UTC | Tue, 26 Jul 2016 20:32:51 UTC |
| + 5797c8550e77764649b38844 | deadlock.delay2test    | stanley      | succeeded (179s elapsed) | Tue, 26 Jul 2016 20:30:13 UTC | Tue, 26 Jul 2016 20:33:12 UTC |
| + 5797c8550e77764649b38847 | deadlock.delay2test    | stanley      | succeeded (201s elapsed) | Tue, 26 Jul 2016 20:30:13 UTC | Tue, 26 Jul 2016 20:33:34 UTC |
| + 5797c8560e77764649b3884a | deadlock.delay2test    | stanley      | succeeded (221s elapsed) | Tue, 26 Jul 2016 20:30:14 UTC | Tue, 26 Jul 2016 20:33:55 UTC |
| + 5797c8560e77764649b3884d | deadlock.delay2test    | stanley      | succeeded (243s elapsed) | Tue, 26 Jul 2016 20:30:14 UTC | Tue, 26 Jul 2016 20:34:17 UTC |
| + 5797c8570e77764649b38850 | deadlock.delay2test    | stanley      | succeeded (264s elapsed) | Tue, 26 Jul 2016 20:30:15 UTC | Tue, 26 Jul 2016 20:34:39 UTC |
| + 5797c8570e77764649b38853 | deadlock.delay2test    | stanley      | succeeded (284s elapsed) | Tue, 26 Jul 2016 20:30:15 UTC | Tue, 26 Jul 2016 20:34:59 UTC |
| + 5797c8580e77764649b38856 | deadlock.delay2test    | stanley      | succeeded (306s elapsed) | Tue, 26 Jul 2016 20:30:16 UTC | Tue, 26 Jul 2016 20:35:22 UTC |
| + 5797c8580e77764649b38859 | deadlock.delay2test    | stanley      | succeeded (68s elapsed)  | Tue, 26 Jul 2016 20:30:16 UTC | Tue, 26 Jul 2016 20:31:24 UTC |
| + 5797c8590e77764649b3885c | deadlock.delay2test    | stanley      | succeeded (326s elapsed) | Tue, 26 Jul 2016 20:30:17 UTC | Tue, 26 Jul 2016 20:35:43 UTC |
| + 5797c8590e77764649b3885f | deadlock.delay2test    | stanley      | succeeded (347s elapsed) | Tue, 26 Jul 2016 20:30:17 UTC | Tue, 26 Jul 2016 20:36:04 UTC |
| + 5797c85a0e77764649b38862 | deadlock.delay2test    | stanley      | succeeded (410s elapsed) | Tue, 26 Jul 2016 20:30:18 UTC | Tue, 26 Jul 2016 20:37:08 UTC |
| + 5797c85a0e77764649b38865 | deadlock.delay2test    | stanley      | succeeded (367s elapsed) | Tue, 26 Jul 2016 20:30:18 UTC | Tue, 26 Jul 2016 20:36:25 UTC |
| + 5797c85b0e77764649b38868 | deadlock.delay2test    | stanley      | succeeded (431s elapsed) | Tue, 26 Jul 2016 20:30:19 UTC | Tue, 26 Jul 2016 20:37:30 UTC |
| + 5797c85c0e77764649b3886b | deadlock.delay2test    | stanley      | succeeded (451s elapsed) | Tue, 26 Jul 2016 20:30:20 UTC | Tue, 26 Jul 2016 20:37:51 UTC |
| + 5797c85c0e77764649b3886e | deadlock.delay2test    | stanley      | succeeded (472s elapsed) | Tue, 26 Jul 2016 20:30:20 UTC | Tue, 26 Jul 2016 20:38:12 UTC |
| + 5797c85d0e77764649b38871 | deadlock.delay2test    | stanley      | succeeded (387s elapsed) | Tue, 26 Jul 2016 20:30:21 UTC | Tue, 26 Jul 2016 20:36:48 UTC |
| + 5797c85e0e77764649b38874 | deadlock.delay2test    | stanley      | succeeded (599s elapsed) | Tue, 26 Jul 2016 20:30:22 UTC | Tue, 26 Jul 2016 20:40:21 UTC |
| + 5797c85e0e77764649b38877 | deadlock.delay2test    | stanley      | succeeded (492s elapsed) | Tue, 26 Jul 2016 20:30:22 UTC | Tue, 26 Jul 2016 20:38:34 UTC |
| + 5797c85f0e77764649b3887a | deadlock.delay2test    | stanley      | succeeded (555s elapsed) | Tue, 26 Jul 2016 20:30:23 UTC | Tue, 26 Jul 2016 20:39:38 UTC |
| + 5797c85f0e77764649b3887d | deadlock.delay2test    | stanley      | succeeded (533s elapsed) | Tue, 26 Jul 2016 20:30:23 UTC | Tue, 26 Jul 2016 20:39:16 UTC |
| + 5797c8600e77764649b38880 | deadlock.delay2test    | stanley      | succeeded (575s elapsed) | Tue, 26 Jul 2016 20:30:24 UTC | Tue, 26 Jul 2016 20:39:59 UTC |
| + 5797c8610e77764649b38883 | deadlock.delay2test    | stanley      | succeeded (639s elapsed) | Tue, 26 Jul 2016 20:30:25 UTC | Tue, 26 Jul 2016 20:41:04 UTC |
| + 5797c8610e77764649b38886 | deadlock.delay2test    | stanley      | succeeded (660s elapsed) | Tue, 26 Jul 2016 20:30:25 UTC | Tue, 26 Jul 2016 20:41:25 UTC |
| + 5797c8620e77764649b3888c | deadlock.delay2test    | stanley      | succeeded (509s elapsed) | Tue, 26 Jul 2016 20:30:26 UTC | Tue, 26 Jul 2016 20:38:55 UTC |
| + 5797c8650e77764649b3889b | deadlock.delay2test    | stanley      | succeeded (614s elapsed) | Tue, 26 Jul 2016 20:30:29 UTC | Tue, 26 Jul 2016 20:40:43 UTC |
+----------------------------+------------------------+--------------+--------------------------+-------------------------------+-------------------------------+

I guess I would expect them to execute pretty much in the order requested

**Comment by Kami** [comment]
Awesome, glad to hear it.

And yes, they should be executed in order as requested. Is the ordering issue the same as one described in #2797?

If so, let's continue there and if not, it's probably the best to open a new issue and we will have a look :)
